From 0de84700648f098c1fbf6b807dee28ec640efe62 Mon Sep 17 00:00:00 2001
From: jmontoyaa <gugli100@gmail.com>
Date: Tue, 29 May 2018 08:27:13 +0200
Subject: [PATCH] Security fix #2532

- Use json_decode/json_encode instead base64
- Add Security::remove_XSSS
---
 main/inc/lib/webservices/Rest.php | 17 ++++-------------
 main/webservices/api/v2.php       |  7 ++++---
 2 files changed, 8 insertions(+), 16 deletions(-)

diff --git a/main/inc/lib/webservices/Rest.php b/main/inc/lib/webservices/Rest.php
index fe5266d9a78..c1e6ae52281 100644
--- a/main/inc/lib/webservices/Rest.php
+++ b/main/inc/lib/webservices/Rest.php
@@ -823,16 +823,9 @@ public function getCourseLearnPaths()
      */
     public static function decodeParams($encoded)
     {
-        $decoded = str_replace(['-', '_', '.'], ['+', '/', '='], $encoded);
-        $mod4 = strlen($decoded) % 4;
+        $decoded = json_decode($encoded);
 
-        if ($mod4) {
-            $decoded .= substr('====', $mod4);
-        }
-
-        $b64Decoded = base64_decode($decoded);
-
-        return unserialize($b64Decoded);
+        return $decoded;
     }
 
     /**
@@ -1319,10 +1312,8 @@ private function encodeParams(array $additionalParams = [])
             'api_key' => $this->apiKey,
             'username' => $this->user->getUsername(),
         ]);
+        $encoded = json_encode($params);
 
-        $strParams = serialize($params);
-        $b64Encoded = base64_encode($strParams);
-
-        return str_replace(['+', '/', '='], ['-', '_', '.'], $b64Encoded);
+        return $encoded;
     }
 }
diff --git a/main/webservices/api/v2.php b/main/webservices/api/v2.php
index 45813ebd8da..f2e7f25728c 100644
--- a/main/webservices/api/v2.php
+++ b/main/webservices/api/v2.php
@@ -6,9 +6,10 @@
 
 if ($hash) {
     $hashParams = Rest::decodeParams($hash);
-
-    foreach ($hashParams as $key => $value) {
-        $_REQUEST[$key] = $value;
+    if (!empty($hashParams)) {
+        foreach ($hashParams as $key => $value) {
+            $_REQUEST[$key] = Security::remove_XSS($value);
+        }
     }
 }
 
